Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp

```diff
@@ -28,26 +28,144 @@
 
 #include "GPUAdapter.h"
 #include "GPUCanvasConfiguration.h"
+#include "RenderBox.h"
+#if PLATFORM(COCOA)
+#include "WebGPUSurfaceCocoa.h"
+#endif
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
+static IntSize getCanvasSizeAsIntSize(const GPUCanvasContext::CanvasType& canvas)
+{
+    return WTF::switchOn(canvas, [](const RefPtr<HTMLCanvasElement>& htmlCanvas) -> IntSize {
+        return { static_cast<int>(htmlCanvas->width()), static_cast<int>(htmlCanvas->height()) };
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> IntSize {
+        return { static_cast<int>(offscreenCanvas->width()), static_cast<int>(offscreenCanvas->height()) };
+    }
+#endif
+    );
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(GPUCanvasContext);
+
+std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
+{
+    return std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
+}
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+    , m_currentConfiguration()
+    , m_surface()
+{
+    suspendIfNeeded();
+}
+
+void GPUCanvasContext::reshape(int width, int height)
+{
+    if (!m_currentConfiguration)
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create(width, height, m_currentConfiguration->convertToBacking());
+    ASSERT(m_surface);
+#else
+    UNUSED_PARAM(width);
+    UNUSED_PARAM(height);
+#endif
+}
+
 GPUCanvasContext::CanvasType GPUCanvasContext::canvas()
 {
-    return { };
+    return htmlCanvas();
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration&& configuration)
 {
+    m_currentConfiguration = WTFMove(configuration);
+
+    auto canvasSize = getCanvasSizeAsIntSize(htmlCanvas());
+    reshape(canvasSize.width(), canvasSize.height());
+}
 
+RefPtr<GPUTexture> GPUCanvasContext::getCurrentTexture()
+{
+    markContextChangedAndNotifyCanvasObservers();
+    return GPUTexture::create(m_surface->currentTexture());
 }
 
 void GPUCanvasContext::unconfigure()
 {
+    m_currentConfiguration.reset();
+    m_surface = nullptr;
 }
 
-RefPtr<GPUTexture> GPUCanvasContext::getCurrentTexture()
+PixelFormat GPUCanvasContext::pixelFormat() const
 {
-    return nullptr;
+    if (!m_surface)
+        return PixelFormat::BGRA8;
+
+    return m_surface->pixelFormat();
+}
+
+DestinationColorSpace GPUCanvasContext::colorSpace() const
+{
+    if (!m_surface)
+        return DestinationColorSpace::SRGB();
+
+    return m_surface->colorSpace();
+}
+
+RefPtr<GraphicsLayerContentsDisplayDelegate> GPUCanvasContext::layerContentsDisplayDelegate()
+{
+    if (!m_surface)
```

Comment:
This if check seems unnecessary. Why not always return m_surface?

Reply:
Fixed.